What is this lab about?

👉 This lab teaches you how to clean messy healthcare data and make it ready for machine learning while keeping patient data safe and private.

🏥 What problems does healthcare data have?

Real medical data is often:

Missing values ❌
Duplicates ❌
Inconsistent (e.g., “M”, “Male”) ❌
Mixed units (kg vs lbs) ❌
Contains private patient info ❌

👉 If you don’t fix these → models give wrong results ⚠️

🔧 What you will learn to do
1. 🧹 Clean the data
Handle missing values
Remove duplicates
Fix inconsistent entries
Detect and handle outliers
2. 🔐 Protect patient privacy
Remove personal data (name, ID, etc.)
Follow rules like:
HIPAA
GDPR

👉 Very important in healthcare

3. ⚙️ Transform data

Make data consistent:

Standardize (same units)
Normalize (scale values)
Encode (convert text → numbers)
4. 🧠 Feature engineering

Create better features like:

BMI (from height & weight)
Time-based features (trends, duration)

👉 Helps model perform better

5. 🤖 Prepare for ML
Encode categorical data
Scale numerical data

👉 So models can understand the data

🪜 Simple Workflow (VERY IMPORTANT)

👉 Always follow this order:

Remove PII (privacy first 🔐)
Clean data (fix errors 🧹)
Handle missing values
Standardize & transform
Create features
Prepare for ML
🎯 Final Goal

👉 Turn this:

Messy, unsafe healthcare data ❌

Into this:

Clean, consistent, privacy-safe dataset ✅
Ready for prediction (e.g., disease risk)

Here’s a **simple beginner-friendly summary** of your dataset 👇

---

## 🧾 What this dataset is

* A **practice healthcare dataset** (fake but realistic)
* Contains **200 patient records**
* Designed to mimic real hospital data (like Electronic Health Records systems)

👉 Important: It is **intentionally messy** so you can learn data cleaning.

---

## ⚠️ Why it’s messy (on purpose)

The dataset includes real-world problems like:

* Missing values (some ages not filled)
* Inconsistent labels (M, Male, F, Female)
* Mixed units (kg vs lbs)
* Different date formats
* Typos in diagnosis (ANX vs ANXITY)
* Duplicate records

👉 This helps you practice **real data preprocessing skills**

---

## 📊 What kind of data it contains

### 1. 👤 Patient Info

* Patient_ID → unique ID
* Age → patient age
* Gender → inconsistent formats
* Ethnicity → categories (but messy)

---

### 2. 📏 Physical Measurements

* Weight → mixed units (kg / lbs ❗)
* Height_cm → height in cm
* Glucose_mg_dL → blood sugar level

👉 Used for health analysis (e.g., diabetes risk)

---

### 3. 🏥 Medical Information

* Diagnosis_Date → different date formats
* Diagnosis_Code → disease labels (with typos)

---

### 4. 🎯 Target Variable

* Risk → what you want to predict

  * 0 = Low risk
  * 1 = High risk

👉 This is your **y (target variable)**

---

### 5. 🔒 Sensitive Data (PII)

* Patient_Name ❌
* EmailID ❌

👉 Must be **removed for privacy (GDPR/HIPAA)**

---

## 🧠 Big Picture (Very Important)

This dataset teaches you how to:

1. Clean messy data 🧹
2. Standardize formats ⚙️
3. Remove sensitive info 🔒
4. Prepare data for ML 🤖

---

## 🧩 One-line memory trick

👉 **“Messy healthcare data → clean → standardize → remove PII → build model”**

---

 👍


In [1]:
# Install the libraries required for this lab
!pip install pandas
!pip install numpy
!pip install scipy
!pip install scikit-learn

In [2]:
# Optional: suppress warnings for cleaner output

import warnings

warnings.filterwarnings('ignore')

Import require Libraries

In [3]:
import pandas as pd   # For data loading, manipulation, cleaning, and saving (DataFrame operations)
import numpy as np    # For numerical operations, array handling, and missing value operations
import re             # For regular expression pattern matching in text processing

from datetime import datetime, timedelta  # For parsing and manipulating date/time values
import random  # For generating random values during data exploration

from scipy import stats  # For statistical functions like z-score for outlier detection
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder
# StandardScaler: standardizes features (mean=0, std=1) for ML algorithms
# MinMaxScaler: scales features to a fixed range (typically 0-1)
# OneHotEncoder: converts categorical features into binary indicator variables

print("All libraries imported successfully!")
print("Ready to begin healthcare data preprocessing.")

All libraries imported successfully!
Ready to begin healthcare data preprocessing.


Step 1: Load & Explore Data

👉 First, load the dataset and take a quick look at it.

🎯 Why this step is important
Understand what data you have
Find problems like:
Missing values
Wrong formats
Duplicates
Outliers

👉 This helps you decide how to clean the data next

In [4]:
# Load the raw healthcare data from CSV file
df = pd.read_csv("/content/Raw_datas.csv")
# Display the first few rows to get an initial sense of the data
print("First 5 rows of the dataset:")
df.head()

First 5 rows of the dataset:


,Patient_ID,Age,Gender,Ethnicity,Weight,Height_cm,Diagnosis_Date,Diagnosis_Code,Glucose_mg_dL,Risk,Patient_Name,EmailID
0,P_new_63,NaN,NaN,asian,NaN,175.0,2021-10-21,NaN,140.0,0,David Jones,kurt31@example.org
1,P016,50.0,M,NaN,NaN,165.0,2020-05-20,DEP,100.0,0,Cristina White,parkerjennifer@example.org
2,P006,25.0,F,Caucasian,NaN,NaN,15/06/2021,OCD,110.0,0,Laura Bates,sean22@example.net
3,P_new_126,34.0,M,asian,70,160.0,2016-10-19,ocd,NaN,1,Rebecca Smith,johnmcclure@example.net
4,P_new_96,25.0,NaN,Asian,70,180.0,2016-02-09,ANXITY,140.0,0,Brittany Wise,nholloway@example.net


Step 2: Understand Data Quality Issues

👉 Real healthcare data usually has 4 common problems:

Missing data → some values are empty
Duplicates → same record appears multiple times
Inconsistent entries → different labels for same thing (M vs Male)
Outliers → unrealistic values (Age = 200)
🎯 Why this step matters

👉 You must find these problems first before cleaning

🧠 Easy way to remember

👉 “4 problems: Missing, Duplicate, Inconsistent, Extreme”

In [6]:
df.shape


(200, 12)

In [7]:
df.dtypes

,0
Patient_ID,object
Age,float64
Gender,object
Ethnicity,object
Weight,object
Height_cm,float64
Diagnosis_Date,object
Diagnosis_Code,object
Glucose_mg_dL,float64
Risk,int64


In [8]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Patient_ID      200 non-null    object 
 1   Age             172 non-null    float64
 2   Gender          162 non-null    object 
 3   Ethnicity       165 non-null    object 
 4   Weight          156 non-null    object 
 5   Height_cm       160 non-null    float64
 6   Diagnosis_Date  190 non-null    object 
 7   Diagnosis_Code  167 non-null    object 
 8   Glucose_mg_dL   158 non-null    float64
 9   Risk            200 non-null    int64  
 10  Patient_Name    200 non-null    object 
 11  EmailID         200 non-null    object 
dtypes: float64(3), int64(1), object(8)
memory usage: 18.9+ KB


Check  the miising value


In [9]:
df.isnull().sum()


,0
Patient_ID,0
Age,28
Gender,38
Ethnicity,35
Weight,44
Height_cm,40
Diagnosis_Date,10
Diagnosis_Code,33
Glucose_mg_dL,42
Risk,0


In [10]:
print((df.isnull().sum() / len(df) * 100).round(2))

Patient_ID         0.0
Age               14.0
Gender            19.0
Ethnicity         17.5
Weight            22.0
Height_cm         20.0
Diagnosis_Date     5.0
Diagnosis_Code    16.5
Glucose_mg_dL     21.0
Risk               0.0
Patient_Name       0.0
EmailID            0.0
dtype: float64


Check duplicate


In [11]:
dup_rows = df.duplicated(keep=False)

In [12]:
dup_rows.sum()

np.int64(0)

In [14]:
if dup_rows.any():
    display(df[dup_rows])

Why this is important
Duplicates can:
Bias your model
Give wrong predictions
Overrepresent some patients

Identify inconsistent categories in data
“Use .unique() to spot messy labels”


In [15]:
df['Gender'].unique()

array([nan, 'M', 'F', 'Male', 'Female', 'Other'], dtype=object)

In [16]:
df['Ethnicity'].unique()

array(['asian', nan, 'Caucasian', 'Asian', 'African', 'Hispanic',
       'caucasian'], dtype=object)

In [17]:
df['Diagnosis_Code'].unique()

array([nan, 'DEP', 'OCD', 'ocd', 'ANXITY', 'ANX'], dtype=object)

Same category written differently can:

Confuse the model
Create wrong groups
Reduce accuracy

Check mixed units (Weight)

In [18]:
df['Weight'].unique()

array([nan, '70', '110', '150lbs', '90', '75', '160lbs', '80'],
      dtype=object)

Check outliers (Age & Glucose)

In [19]:
df[['Age', 'Glucose_mg_dL']].describe()

,Age,Glucose_mg_dL
count,172.000000,158.000000
mean,70.087209,220.981013
std,62.340026,177.443324
min,25.000000,85.000000
25%,34.000000,90.000000
50%,45.000000,110.000000
75%,60.000000,500.000000
max,200.000000,500.000000


Shows stats like:

min, max, mean
👉 Helps find unrealistic values (e.g., Age = 200)

Check date format issues

In [20]:
df['Diagnosis_Date'].sample(10)

,Diagnosis_Date
133,2018-10-03
157,19/05/2016
10,2021-03-10
134,2025-05-20
19,05/05/2017
117,18/07/2020
32,2020-02-03
81,12/12/2019
22,16/06/2021
152,2021-04-26


Finds mixed formats like:

YYYY-MM-DD
DD/MM/YYYY ❗

Check target balance (Risk)

In [21]:
df['Risk'].value_counts()

,count
Risk,
0,107
1,93


Shows percentage

👉 Detects class imbalance

⚠️ Detecting Outliers (Important Step)

👉 Outliers = unusual or extreme values (can harm your model)

Detecting Outliers (Important Step)

👉 Outliers = unusual or extreme values (can harm your model)

🧠 Two methods used
1. 📦 IQR Method (most common)
Uses:
Q1 (25%)
Q3 (75%)
Formula:
IQR = Q3 − Q1
Outlier if:
Value < Q1 − 1.5×IQR
Value > Q3 + 1.5×IQR

👉 Best when data is not normally distributed

📊 Z-Score Method
Formula:
Z = (value − mean) / std
Outlier if:
|Z| > 3

👉 Best for normal (bell-shaped) data

In [22]:
def iqr_outliers(series):
    """Return outliers in a pandas Series using the IQR method."""
    q1, q3 = series.quantile([0.25, 0.75])
    iqr = q3 - q1
    return series[(series < q1 - 1.5 * iqr) | (series > q3 + 1.5 * iqr)]

# Function to print outliers for a column
def print_outliers(df, column):
    outliers = iqr_outliers(df[column].dropna())
    print(f"{column} outliers (IQR method):")
    print(f"Found {len(outliers)} outliers")
    if len(outliers) > 0:
        print(f"Outlier values: {outliers.unique()}\n")

# Detect outliers
print_outliers(df, 'Age')
print_outliers(df, 'Glucose_mg_dL')

Age outliers (IQR method):
Found 31 outliers
Outlier values: [200.]

Glucose_mg_dL outliers (IQR method):
Found 0 outliers


When handling healthcare data, protecting privacy is crucial. Personally Identifiable Information (PII) — such as names, contact info, patient IDs, and sensitive dates — must be removed or anonymized before analysis or sharing to comply with regulations like HIPAA and GDPR. Always create a clean copy of the dataset to preserve the original while removing PII.

In [23]:
# Create a working copy - keep original data untouched for reference
df_clean = df.copy()
print("Working copy created. Original data preserved.")

Working copy created. Original data preserved.


In [24]:
# Create a working copy - keep original data untouched for reference
df_clean = df.copy()
print("Working copy created. Original data preserved.")

Working copy created. Original data preserved.


In [25]:
# Identify and remove PII columns
pii_columns = ['Patient_Name', 'EmailID']
print(f"Removing PII columns: {pii_columns}")

df_clean = df_clean.drop(columns=pii_columns, errors='ignore')

print("\nColumns after removing PII:")
print(df_clean.columns.tolist())
print(f"\nReduced from {len(df.columns)} to {len(df_clean.columns)} columns")

Removing PII columns: ['Patient_Name', 'EmailID']

Columns after removing PII:
['Patient_ID', 'Age', 'Gender', 'Ethnicity', 'Weight', 'Height_cm', 'Diagnosis_Date', 'Diagnosis_Code', 'Glucose_mg_dL', 'Risk']

Reduced from 12 to 10 columns


Duplicate rows can arise from errors or merges, leading to biased analysis, inflated dataset size, or data leakage. Identify and remove exact duplicates, keeping only the first occurrence.

In [26]:
# Count duplicates before removal
duplicates_before = df_clean.duplicated().sum()
rows_before = len(df_clean)

# Remove exact duplicate rows (keep first occurrence)
df_clean = df_clean.drop_duplicates(keep='first')

# Report results
rows_after = len(df_clean)
print(f"Rows before deduplication: {rows_before}")
print(f"Duplicate rows found: {duplicates_before}")
print(f"Rows after deduplication: {rows_after}")
print(f"Rows removed: {rows_before - rows_after}")

Rows before deduplication: 200
Duplicate rows found: 2
Rows after deduplication: 198
Rows removed: 2


Standardize inconsistent categorical variables

Inconsistent categorical data is common in healthcare due to:
- Multiple data entry personnel with different conventions
- Data merging from different systems
- Typos and abbreviations

You need to standardize:
- **Gender**: Convert M/Male/m → 'Male', F/Female/f → 'Female'
- **Ethnicity**: Standardize capitalization (asian/Asian/ASIAN → 'Asian')
- **Diagnosis_Code**: Fix typos and standardize (OCD/ocd → 'OCD', ANXITY → 'ANX')

This ensures categorical data is **clean, consistent, and machine-readable**.

In [27]:
# Standardize Gender column
print("Before standardization - Gender unique values:")
print(df_clean['Gender'].value_counts(dropna=False))

# Convert to lowercase and strip whitespace for consistent matching
df_clean['Gender'] = df_clean['Gender'].astype(str).str.strip().str.lower()

# Define mapping for known variations
gender_map = {
    'male': 'Male', 'm': 'Male',
    'female': 'Female', 'f': 'Female',
    'other': 'Other',
    'nan': np.nan, 'none': np.nan
}

# Apply mapping
df_clean['Gender'] = df_clean['Gender'].replace({'nan': np.nan})
df_clean['Gender'] = df_clean['Gender'].map(
    lambda x: gender_map.get(x, x.capitalize() if pd.notna(x) else x)
)

print("\nAfter standardization - Gender unique values:")
print(df_clean['Gender'].value_counts(dropna=False))

Before standardization - Gender unique values:
Gender
M         46
F         39
NaN       38
Female    35
Male      33
Other      7
Name: count, dtype: int64

After standardization - Gender unique values:
Gender
Male      79
Female    74
NaN       38
Other      7
Name: count, dtype: int64


In [29]:
# Standardize Ethnicity column
print("Before standardization - Ethnicity unique values:")
print(df_clean['Ethnicity'].value_counts(dropna=False))

# Clean and capitalize values
df_clean['Ethnicity'] = df_clean['Ethnicity'].astype(str).str.strip().replace({'nan': np.nan})
df_clean['Ethnicity'] = df_clean['Ethnicity'].str.capitalize()

print("\nAfter standardization - Ethnicity unique values:")
print(df_clean['Ethnicity'].value_counts(dropna=False))

Before standardization - Ethnicity unique values:
Ethnicity
Asian        75
NaN          35
African      32
Hispanic     31
Caucasian    25
Name: count, dtype: int64

After standardization - Ethnicity unique values:
Ethnicity
Asian        75
NaN          35
African      32
Hispanic     31
Caucasian    25
Name: count, dtype: int64


df_clean['Ethnicity'] selects the column named Ethnicity.
.value_counts(dropna=False) counts how many times each value appears, including missing values (NaN).
This is useful to see what your data looks like before cleaning.


.astype(str) ensures that every value is a string (sometimes numbers or NaNs are stored differently).
.str.strip() removes spaces at the start or end of the string (" African " → "African").
This avoids tiny differences that make the same value appear multiple times.

Sometimes missing values are stored as the string "nan".
.replace({'nan': np.nan}) turns these into actual missing values (np.nan), which pandas understands.


.str.capitalize() makes the first letter uppercase and the rest lowercase.
Example: "asian" → "Asian", "african american" → "African american".

This shows the cleaned column and helps verify the standardization worked.

In [30]:
# Standardize Diagnosis_Code column
print("Before standardization - Diagnosis_Code unique values:")
print(df_clean['Diagnosis_Code'].value_counts(dropna=False))

# Convert to uppercase and fix common typos
df_clean['Diagnosis_Code'] = df_clean['Diagnosis_Code'].astype(str).str.strip().str.upper()
df_clean['Diagnosis_Code'] = df_clean['Diagnosis_Code'].replace({
    'NAN': np.nan,
    'ANXITY': 'ANX',  # Fix typo
    'OCD.': 'OCD'      # Remove trailing period
})

print("\nAfter standardization - Diagnosis_Code unique values:")
print(df_clean['Diagnosis_Code'].value_counts(dropna=False))

Before standardization - Diagnosis_Code unique values:
Diagnosis_Code
OCD       40
DEP       39
ocd       35
NaN       32
ANX       31
ANXITY    21
Name: count, dtype: int64

After standardization - Diagnosis_Code unique values:
Diagnosis_Code
OCD    75
ANX    52
DEP    39
NaN    32
Name: count, dtype: int64


Filling this  categorical value  with Unknown

.fillna('Unknown') is a pandas method that replaces all missing values (NaN) with a specified value ('Unknown' in this case).
This is common for categorical data (like Gender or Ethnicity), because missing values can cause problems in analysis or machine learning.
Here, we do this for three columns: Gender, Ethnicity, and Diagnosis_Code.

Example:
Before: ['Male', NaN, 'Female', NaN]
After: ['Male', 'Unknown', 'Female', 'Unknown']

In [31]:
df_clean['Gender'] = df_clean['Gender'].fillna('Unknown')
df_clean['Ethnicity'] = df_clean['Ethnicity'].fillna('Unknown')
df_clean['Diagnosis_Code'] = df_clean['Diagnosis_Code'].fillna('Unknown')

check the result

In [32]:
print("\nFinal categorical value counts:")
print("\nGender:")
print(df_clean['Gender'].value_counts())


Final categorical value counts:

Gender:
Gender
Male       79
Female     74
Unknown    38
Other       7
Name: count, dtype: int64


Feature engineering

Focuses on normalizing units and creating a BMI feature:

Healthcare data may mix units (kg vs lbs, cm vs m), so first convert all weights to kg and heights to meters.
Then calculate BMI using:
BMI=Weight (kg)Height (m)2
BMI=
Height (m)
2
Weight (kg)
	​

BMI is useful for classifying patients:
Underweight: < 18.5
Normal: 18.5–24.9
Overweight: 25–29.9
Obese: ≥ 30

Define a function to standardize messy input
Handles different formats, units, or missing values.
Example: weight_to_kg() converts numbers and strings like '150 lbs' into kilograms.
Handle missing values
pd.isna(x) keeps NaN intact so you don’t accidentally change missing data.
Detect and handle special cases
Strings with 'lb' → convert to kg.
Numbers → assume already in kg.
Fallback to a safe default
If conversion fails (unexpected format), return NaN.
Apply function to a column
df_clean['Weight'].apply(weight_to_kg) applies your function to every row in the column.
Verify the transformation
display(df_clean[['Weight', 'Weight_kg']].head(10)) shows examples to make sure the function worked.

💡 Takeaway for beginners:
Whenever your dataset has mixed formats or units, create a small, clear function to standardize it, then apply it column-wise. Always handle missing or unexpected data safely!

In [33]:


# Function to convert weight to kilograms
def weight_to_kg(x):
    if pd.isna(x):  # Keep missing values as NaN
        return np.nan

    # If it's already a number, assume kg
    if isinstance(x, (int, float)):
        return float(x)

    # Convert string to lowercase and remove spaces
    s = str(x).strip().lower()

    # Convert lbs to kg
    if 'lb' in s:
        number = float(re.findall(r'\d+\.?\d*', s)[0])
        return round(number * 0.4536, 2)

    # Otherwise, try converting directly to float (assume kg)
    try:
        return float(s)
    except:
        return np.nan

# Apply function to the Weight column
df_clean['Weight_kg'] = df_clean['Weight'].apply(weight_to_kg)

# Show some examples
print("Weight conversion examples:")
display(df_clean[['Weight', 'Weight_kg']].head(10))

Weight conversion examples:


,Weight,Weight_kg
0,NaN,NaN
1,NaN,NaN
2,NaN,NaN
3,70,70.00
4,70,70.00
5,110,110.00
6,NaN,NaN
7,NaN,NaN
8,150lbs,68.04
9,150lbs,68.04


In [34]:
# Convert height from cm to meters
df_clean['Height_cm'] = pd.to_numeric(df_clean['Height_cm'], errors='coerce')
df_clean['Height_m'] = df_clean['Height_cm'] / 100.0

print("Height converted from cm to meters")

Height converted from cm to meters


pd.to_numeric() tries to turn every value in the column into a number.
errors='coerce' means that if a value cannot be converted, it becomes NaN instead of crashing.
This is useful when your data might have text, missing values, or typos


Heights in your data are in centimeters.
Divide by 100 to convert to meters (since 1 m = 100 cm).
This creates a new column Height_m while keeping the original Height_cm column.

Define a calculation function
calculate_bmi(weight, height) takes two inputs and returns a BMI.
Always check for missing or invalid values (NaN or height = 0) to avoid errors.
Apply the function row-wise
df.apply(..., axis=1) applies a function to each row in the DataFrame.
lambda row: calculate_bmi(...) passes the row’s weight and height to the function.
Create a new column
Store the result in a new column BMI for further analysis.
This is common when you derive clinically or mathematically meaningful features from raw data.
Inspect results
Use display(df[['Weight', 'Weight_kg', 'Height_cm', 'Height_m', 'BMI']].head(10)) to check a sample of the new feature.

In [35]:
# Calculate BMI (Body Mass Index)
def calculate_bmi(weight, height):
    if pd.notna(weight) and pd.notna(height) and height > 0:
        return round(weight / (height ** 2), 2)
    else:
        return np.nan

df_clean['BMI'] = df_clean.apply(lambda row: calculate_bmi(row['Weight_kg'], row['Height_m']), axis=1)

print("BMI calculated successfully")
display(df_clean[['Weight', 'Weight_kg', 'Height_cm', 'Height_m', 'BMI']].head(10))

BMI calculated successfully


,Weight,Weight_kg,Height_cm,Height_m,BMI
0,NaN,NaN,175.0,1.75,NaN
1,NaN,NaN,165.0,1.65,NaN
2,NaN,NaN,NaN,NaN,NaN
3,70,70.00,160.0,1.60,27.34
4,70,70.00,180.0,1.80,21.60
5,110,110.00,180.0,1.80,33.95
6,NaN,NaN,175.0,1.75,NaN
7,NaN,NaN,175.0,1.75,NaN
8,150lbs,68.04,NaN,NaN,NaN
9,150lbs,68.04,NaN,NaN,NaN


Handling missing values in healthcare data is essential because gaps can come from skipped tests, errors, equipment issues, or privacy restrictions.

Median imputation is preferred over the mean because it:

Is robust to outliers and extreme values.
Represents the middle value, unaffected by extremes.
Preserves the data’s distribution better.
Is simple and fast to compute.

In [36]:
# List of numeric columns to impute
numeric_cols = ['Age', 'Weight_kg', 'Height_cm', 'BMI', 'Glucose_mg_dL']

# Check missing values before imputation
print("Missing values before imputation:")
print(df_clean[numeric_cols].isnull().sum())

# Median imputation
for col in numeric_cols:
    df_clean[col].fillna(df_clean[col].median(), inplace=True)

# Verify missing values are filled
print("\nMissing values after imputation:")
print(df_clean[numeric_cols].isnull().sum())
print("\nAll numeric missing values successfully imputed!")

Missing values before imputation:
Age              28
Weight_kg        43
Height_cm        39
BMI              75
Glucose_mg_dL    42
dtype: int64

Missing values after imputation:
Age              0
Weight_kg        0
Height_cm        0
BMI              0
Glucose_mg_dL    0
dtype: int64

All numeric missing values successfully imputed!


Checks how many NaN values are in each column.

Loop through multiple columns
Makes it efficient to impute many numeric columns at once.
Verify imputation
Recheck missing values to ensure all gaps are filled.

Parsing dates and creating temporal features helps extract useful insights from healthcare data:

Diagnosis year – shows trends or changes over time.
Time since diagnosis – helps track disease progression.
Seasonal patterns – some conditions vary by time of year.

Convert strings to datetime objects
errors='coerce' ensures any unparseable date becomes NaT instead of causing an error.
dayfirst=True handles formats like DD/MM/YYYY.
Create a new column
Store the cleaned/parsed dates in a separate column (Diagnosis_Date_parsed) to keep original data intact.
Verify results
Count parsed vs failed dates using .notna().sum() and .isna().sum().
Display a few examples to visually confirm parsing worked.

In [37]:
# Parse diagnosis dates with mixed formats
df_clean['Diagnosis_Date_parsed'] = pd.to_datetime(
    df_clean['Diagnosis_Date'],
    errors='coerce',   # Unparseable dates become NaT
    dayfirst=True      # Treat ambiguous dates as day/month/year
)

# Check how many dates were parsed successfully
print("Date parsing results:")
print("Successfully parsed:", df_clean['Diagnosis_Date_parsed'].notna().sum())
print("Failed to parse:", df_clean['Diagnosis_Date_parsed'].isna().sum())

# Display a sample of original vs parsed dates
display(df_clean[['Diagnosis_Date', 'Diagnosis_Date_parsed']].head(10))

Date parsing results:
Successfully parsed: 87
Failed to parse: 111


,Diagnosis_Date,Diagnosis_Date_parsed
0,2021-10-21,2021-10-21
1,2020-05-20,2020-05-20
2,15/06/2021,NaT
3,2016-10-19,2016-10-19
4,2016-02-09,2016-02-09
5,2024-05-13,2024-05-13
6,2017-12-31,2017-12-31
7,2022-12-15,2022-12-15
8,2024-03-04,2024-03-04
9,2022-10-10,2022-10-10


In [38]:
# Extract year from diagnosis date
df_clean['Diagnosis_Year'] = df_clean['Diagnosis_Date_parsed'].dt.year

# Use the most recent diagnosis date as reference, or today if missing
ref_date = df_clean['Diagnosis_Date_parsed'].max()
if pd.isna(ref_date):
    ref_date = pd.to_datetime("today")

# Calculate days since diagnosis
df_clean['Days_Since_Diagnosis'] = (ref_date - df_clean['Diagnosis_Date_parsed']).dt.days

# Display results
print(f"Reference date for calculating days since diagnosis: {ref_date.date()}")
display(df_clean[['Diagnosis_Date_parsed', 'Diagnosis_Year', 'Days_Since_Diagnosis']].head(10))

Reference date for calculating days since diagnosis: 2025-06-27


,Diagnosis_Date_parsed,Diagnosis_Year,Days_Since_Diagnosis
0,2021-10-21,2021.0,1345.0
1,2020-05-20,2020.0,1864.0
2,NaT,NaN,NaN
3,2016-10-19,2016.0,3173.0
4,2016-02-09,2016.0,3426.0
5,2024-05-13,2024.0,410.0
6,2017-12-31,2017.0,2735.0
7,2022-12-15,2022.0,925.0
8,2024-03-04,2024.0,480.0
9,2022-10-10,2022.0,991.0


Take a reference date (ref_date) – usually the most recent date or today.
Subtract each patient’s diagnosis date from it → gives a timedelta.
Convert the timedelta to days using .dt.days.
Store it in a new column for analysis or modeling.

This is very useful in healthcare for features like:

Time since diagnosis
Age of a condition
Tracking disease progression


The code calculates the number of days “in between” the most recent reference date (ref_date) and each patient’s previous diagnosis date.
It tells you how long ago each patient was diagnosed relative to that reference.

For example, if ref_date is March 1, 2023 and a patient was diagnosed on January 1, 2023, then

Encoding categorical column


Encoding categorical variables means converting text categories into numbers so machine learning models can understand them.

One-hot encoding creates a separate 0/1 column for each category.

Example: Diagnosis_Code = ['DEP', 'OCD', 'ANX'] → columns: Diagnosis_Code_DEP, Diagnosis_Code_OCD, Diagnosis_Code_ANX
A patient with 'OCD' would have [0, 1, 0].

Why use one-hot encoding?

Treats all categories equally (no order assumed)
Works with any ML algorithm
Prevents models from misinterpreting category numbers as having value relationships

Note: Label encoding (1, 2, 3…) is only for categories with a natural order.

In practice: Apply one-hot encoding to columns like Gender, Ethnicity, and Diagnosis_Code

In [39]:
# Columns to encode
categorical_cols = ['Diagnosis_Code', 'Gender', 'Ethnicity']

# Check columns before encoding
print("Columns before encoding:", df_clean.columns.tolist())
print("Total columns:", len(df_clean.columns))

# Apply one-hot encoding
df_final = pd.get_dummies(df_clean, columns=categorical_cols, drop_first=False)

# Check columns after encoding
print("\nColumns after encoding:", df_final.columns.tolist())
print("Total columns:", len(df_final.columns))
print("New encoded columns added:", len(df_final.columns) - len(df_clean.columns))

Columns before encoding: ['Patient_ID', 'Age', 'Gender', 'Ethnicity', 'Weight', 'Height_cm', 'Diagnosis_Date', 'Diagnosis_Code', 'Glucose_mg_dL', 'Risk', 'Weight_kg', 'Height_m', 'BMI', 'Diagnosis_Date_parsed', 'Diagnosis_Year', 'Days_Since_Diagnosis']
Total columns: 16

Columns after encoding: ['Patient_ID', 'Age', 'Weight', 'Height_cm', 'Diagnosis_Date', 'Glucose_mg_dL', 'Risk', 'Weight_kg', 'Height_m', 'BMI', 'Diagnosis_Date_parsed', 'Diagnosis_Year', 'Days_Since_Diagnosis', 'Diagnosis_Code_ANX', 'Diagnosis_Code_DEP', 'Diagnosis_Code_OCD', 'Diagnosis_Code_Unknown', 'Gender_Female', 'Gender_Male', 'Gender_Other', 'Gender_Unknown', 'Ethnicity_African', 'Ethnicity_Asian', 'Ethnicity_Caucasian', 'Ethnicity_Hispanic', 'Ethnicity_Unknown']
Total columns: 26
New encoded columns added: 10


Machine learning models cannot use text directly, so you need to convert categories to numbers.
One-hot encoding is the safest choice for non-ordinal categories.
Always verify new columns after encoding to make sure everything looks correct.

Scalling the feature

Scaling numeric features is important because many machine learning algorithms are sensitive to the size of the numbers.

Example: Age ranges 0–100, Glucose ranges 70–500. Without scaling, larger numbers can dominate the model.
Algorithms affected: Logistic Regression, SVM, KNN, Neural Networks, K-Means
Algorithms not affected: Tree-based models like Decision Trees or Random Forests

StandardScaler (Z-score normalization) transforms numbers using:

z=x−meanstd
z=
std
x−mean
	​

Resulting data has mean = 0 and standard deviation = 1.
Best for algorithms that assume normally distributed data.
Values typically range from about -3 to +3.

In [40]:
from sklearn.preprocessing import StandardScaler

# 1. Define numeric columns to scale
numeric_cols = ['Age', 'Weight_kg', 'Height_cm', 'BMI', 'Glucose_mg_dL', 'Days_Since_Diagnosis']
numeric_cols = [col for col in numeric_cols if col in df_final.columns]  # Keep only existing columns

print(f"Scaling {len(numeric_cols)} numeric features:", numeric_cols)

# 2. Fill any remaining missing values with median
df_final[numeric_cols] = df_final[numeric_cols].fillna(df_final[numeric_cols].median())

print("Verified no missing values before scaling:")
print(df_final[numeric_cols].isna().sum())

# 3. Apply StandardScaler
scaler = StandardScaler()
scaled_cols = [col + '_scaled' for col in numeric_cols]
df_final[scaled_cols] = scaler.fit_transform(df_final[numeric_cols])

print("Scaling complete!")

# 4. Check statistics of scaled features
display(df_final[scaled_cols].describe())

# 5. Compare original vs scaled example columns
display(df_final[['Age', 'Age_scaled', 'Glucose_mg_dL', 'Glucose_mg_dL_scaled']].head(10))

Scaling 6 numeric features: ['Age', 'Weight_kg', 'Height_cm', 'BMI', 'Glucose_mg_dL', 'Days_Since_Diagnosis']
Verified no missing values before scaling:
Age                     0
Weight_kg               0
Height_cm               0
BMI                     0
Glucose_mg_dL           0
Days_Since_Diagnosis    0
dtype: int64
Scaling complete!


,Age_scaled,Weight_kg_scaled,Height_cm_scaled,BMI_scaled,Glucose_mg_dL_scaled,Days_Since_Diagnosis_scaled
count,1.980000e+02,1.980000e+02,1.980000e+02,1.980000e+02,198.000000,1.980000e+02
mean,7.850062e-17,6.728624e-18,9.644362e-16,-6.369764e-16,0.000000,1.794300e-16
std,1.002535e+00,1.002535e+00,1.002535e+00,1.002535e+00,1.002535,1.002535e+00
min,-7.156602e-01,-8.994916e-01,-1.154814e+00,-1.558938e+00,-0.692822,-2.763449e+00
25%,-5.618235e-01,-7.654342e-01,-4.555688e-01,-4.181248e-01,-0.540469,2.006038e-02
50%,-3.738008e-01,-4.234511e-01,-4.555688e-01,-2.198113e-01,-0.540469,2.006038e-02
75%,-1.174063e-01,6.024982e-01,9.429215e-01,1.563709e-01,-0.357645,2.006038e-02
max,2.275609e+00,1.970431e+00,1.642167e+00,2.932759e+00,1.836241,3.071343e+00


,Age,Age_scaled,Glucose_mg_dL,Glucose_mg_dL_scaled
0,45.0,-0.373801,140.0,-0.357645
1,50.0,-0.288336,100.0,-0.601410
2,25.0,-0.715660,110.0,-0.540469
3,34.0,-0.561823,110.0,-0.540469
4,25.0,-0.715660,140.0,-0.357645
5,25.0,-0.715660,110.0,-0.540469
6,45.0,-0.373801,90.0,-0.662352
7,60.0,-0.117406,110.0,-0.540469
8,45.0,-0.373801,140.0,-0.357645
9,45.0,-0.373801,90.0,-0.662352


Always scale numeric features when using algorithms sensitive to feature magnitude (like Logistic Regression, SVM, KNN, Neural Networks).
Keep original values if you want to compare or debug later.
Fill missing values first, because scaling cannot handle NaNs.

Saving the cleaned dataset means writing your fully processed data to a CSV file.

After this, the dataset is ready for:
Statistical analysis
Exploratory data analysis (EDA)
Machine learning model training



Saving the data
Step 1: Choose a file name and path.
Step 2: Save the cleaned/processed DataFrame to a CSV.
Step 3: Print details to confirm the save and check dataset shape.
Step 4: Optionally, show the current directory to locate the file easily.

In [41]:
# Save the cleaned dataset
df_final.to_csv("health_data_cleaned.csv", index=False)

# Confirm save and show dataset shapes
print("✓ Cleaned dataset saved as: health_data_cleaned.csv")
print(f"Final dataset: {df_final.shape[0]} rows × {df_final.shape[1]} columns")
print(f"Original dataset: {df.shape[0]} rows × {df.shape[1]} columns")

✓ Cleaned dataset saved as: health_data_cleaned.csv
Final dataset: 198 rows × 32 columns
Original dataset: 200 rows × 12 columns


## Step 13: Evaluate data cleaning results

Let's compare the raw and cleaned datasets to verify preprocessing was successful.

Compare column structures
Shows which columns exist in the raw, cleaned, and final encoded datasets.
Compare missing values
Displays missing values per column and total missing values for raw vs cleaned datasets.
Compare statistical summaries
Uses .describe() to see how numeric distributions changed after cleaning.
Compare categorical standardization
Example: Gender shows how inconsistent values in raw data were cleaned into consistent categories.

In [42]:
# =========================
# COLUMN COMPARISON
# =========================
print("\nCOLUMN COMPARISON")
print("Raw data columns:", df.columns.tolist())
print("Cleaned data columns:", df_clean.columns.tolist())
print("Final encoded data columns:", df_final.columns.tolist())

# =========================
# MISSING VALUES COMPARISON
# =========================
print("\nMISSING VALUES COMPARISON")
print("Raw data missing values:\n", df.isna().sum())
print("Total missing values in raw data:", df.isna().sum().sum())

print("\nCleaned data missing values:\n", df_clean.isna().sum())
print("Total missing values in cleaned data:", df_clean.isna().sum().sum())

# =========================
# STATISTICAL SUMMARY COMPARISON
# =========================
print("\nSTATISTICAL SUMMARY COMPARISON")
print("Raw data summary:")
display(df.describe())

print("Cleaned data summary:")
display(df_clean.describe())

# =========================
# CATEGORICAL STANDARDIZATION - GENDER EXAMPLE
# =========================
print("\nCATEGORICAL STANDARDIZATION - GENDER EXAMPLE")
print("Raw Gender value counts:\n", df['Gender'].value_counts(dropna=False))
print("\nCleaned Gender value counts:\n", df_clean['Gender'].value_counts(dropna=False))
print("\nSuccessfully standardized from multiple variations to consistent categories!")


COLUMN COMPARISON
Raw data columns: ['Patient_ID', 'Age', 'Gender', 'Ethnicity', 'Weight', 'Height_cm', 'Diagnosis_Date', 'Diagnosis_Code', 'Glucose_mg_dL', 'Risk', 'Patient_Name', 'EmailID']
Cleaned data columns: ['Patient_ID', 'Age', 'Gender', 'Ethnicity', 'Weight', 'Height_cm', 'Diagnosis_Date', 'Diagnosis_Code', 'Glucose_mg_dL', 'Risk', 'Weight_kg', 'Height_m', 'BMI', 'Diagnosis_Date_parsed', 'Diagnosis_Year', 'Days_Since_Diagnosis']
Final encoded data columns: ['Patient_ID', 'Age', 'Weight', 'Height_cm', 'Diagnosis_Date', 'Glucose_mg_dL', 'Risk', 'Weight_kg', 'Height_m', 'BMI', 'Diagnosis_Date_parsed', 'Diagnosis_Year', 'Days_Since_Diagnosis', 'Diagnosis_Code_ANX', 'Diagnosis_Code_DEP', 'Diagnosis_Code_OCD', 'Diagnosis_Code_Unknown', 'Gender_Female', 'Gender_Male', 'Gender_Other', 'Gender_Unknown', 'Ethnicity_African', 'Ethnicity_Asian', 'Ethnicity_Caucasian', 'Ethnicity_Hispanic', 'Ethnicity_Unknown', 'Age_scaled', 'Weight_kg_scaled', 'Height_cm_scaled', 'BMI_scaled', 'Glucose_m

,Age,Height_cm,Glucose_mg_dL,Risk
count,172.000000,160.000000,158.000000,200.000000
mean,70.087209,169.031250,220.981013,0.465000
std,62.340026,7.780335,177.443324,0.500025
min,25.000000,160.000000,85.000000,0.000000
25%,34.000000,160.000000,90.000000,0.000000
50%,45.000000,165.000000,110.000000,0.000000
75%,60.000000,175.000000,500.000000,1.000000
max,200.000000,180.000000,500.000000,1.000000


Cleaned data summary:


,Age,Height_cm,Glucose_mg_dL,Risk,Weight_kg,Height_m,BMI,Diagnosis_Date_parsed,Diagnosis_Year,Days_Since_Diagnosis
count,198.000000,198.000000,198.000000,198.000000,198.000000,159.000000,198.000000,87,87.000000,87.000000
mean,66.868687,168.257576,198.686869,0.469697,81.191111,1.690566,28.625152,2020-12-13 14:53:47.586206976,2020.482759,1656.379310
min,25.000000,160.000000,85.000000,0.000000,68.040000,1.600000,21.000000,2015-10-28 00:00:00,2015.000000,0.000000
25%,34.000000,165.000000,110.000000,0.000000,70.000000,1.600000,26.580000,2019-01-14 12:00:00,2019.000000,944.000000
50%,45.000000,165.000000,110.000000,0.000000,75.000000,1.650000,27.550000,2020-11-16 00:00:00,2020.000000,1684.000000
75%,60.000000,175.000000,140.000000,1.000000,90.000000,1.750000,29.390000,2022-11-26 00:00:00,2022.000000,2355.500000
max,200.000000,180.000000,500.000000,1.000000,110.000000,1.800000,42.970000,2025-06-27 00:00:00,2025.000000,3530.000000
std,58.651885,7.168694,164.508321,0.500346,14.657666,0.077983,4.903646,NaN,2.573849,917.743203



CATEGORICAL STANDARDIZATION - GENDER EXAMPLE
Raw Gender value counts:
 Gender
M         46
F         40
NaN       38
Female    36
Male      33
Other      7
Name: count, dtype: int64

Cleaned Gender value counts:
 Gender
Male       79
Female     74
Unknown    38
Other       7
Name: count, dtype: int64

Successfully standardized from multiple variations to consistent categories!
